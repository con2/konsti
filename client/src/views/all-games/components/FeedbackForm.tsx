import React, { FC, ReactElement, ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { postFeedback } from 'services/feedbackServices';
import { Game } from 'typings/game.typings';

export interface Props {
  game: Game;
}

export const FeedbackForm: FC<Props> = (props: Props): ReactElement => {
  const { game } = props;

  const [submitting, setSubmitting] = React.useState<boolean>(false);
  const [feedbackValue, setFeedbackValue] = React.useState<string>('');
  const [feedbackSent, setFeedbackSent] = React.useState<boolean>(false);

  const { t } = useTranslation();

  // Hide / show clicked
  const sendFeedbackEvent = async (): Promise<void> => {
    setSubmitting(true);

    const feedbackData = {
      gameId: game.gameId,
      feedback: feedbackValue,
    };

    try {
      await postFeedback(feedbackData);
    } catch (error) {
      console.log(`postFeedback error:`, error);
    }
    setFeedbackSent(true);
    setSubmitting(false);
  };

  const handleFeedbackChange = (
    event: ChangeEvent<HTMLTextAreaElement>
  ): void => {
    setFeedbackValue(event.target.value);
  };

  return (
    <div className='feedback-form'>
      <p className='bold'>{t('feedbackTitle')}</p>
      <p>{t('feedbackInstruction')}</p>

      {!feedbackSent && (
        <>
          <FeedbackTextarea
            value={feedbackValue}
            onChange={handleFeedbackChange}
            rows={4}
          />

          <button
            disabled={submitting}
            onClick={async () => await sendFeedbackEvent()}
          >
            {t('button.sendFeedback')}
          </button>
        </>
      )}

      {feedbackSent && (
        <SuccessMessage>{t('button.feedbackSent')}</SuccessMessage>
      )}
    </div>
  );
};

const FeedbackTextarea = styled.textarea`
  width: 95%;
  border: 1px solid black;
  resize: none;
  overflow: auto;
`;

const SuccessMessage = styled.p`
  color: ${(props) => props.theme.success};
`;
