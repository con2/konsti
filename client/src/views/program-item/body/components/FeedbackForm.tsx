import { ReactElement, ChangeEvent, useState } from "react";
import { useTranslation } from "react-i18next";
import styled from "styled-components";
import { postFeedback } from "client/services/feedbackServices";
import { ProgramItem } from "shared/types/models/programItem";
import { Button, ButtonStyle } from "client/components/Button";
import { TextArea } from "client/components/TextArea";

interface Props {
  programItem: ProgramItem;
}

export const FeedbackForm = ({ programItem }: Props): ReactElement => {
  const [submitting, setSubmitting] = useState<boolean>(false);
  const [feedbackValue, setFeedbackValue] = useState<string>("");
  const [feedbackSent, setFeedbackSent] = useState<boolean>(false);

  const { t } = useTranslation();

  // Hide / show clicked
  const sendFeedbackEvent = async (): Promise<void> => {
    if (feedbackValue.length === 0) {
      return;
    }
    setSubmitting(true);

    try {
      await postFeedback(programItem.programItemId, feedbackValue);
    } catch (error) {
      console.log(`postFeedback error:`, error); // eslint-disable-line no-console
    }
    setFeedbackSent(true);
    setSubmitting(false);
  };

  const handleFeedbackChange = (
    event: ChangeEvent<HTMLTextAreaElement>,
  ): void => {
    setFeedbackValue(event.target.value);
  };

  return (
    <div>
      <Title>{t("feedbackTitle")}</Title>
      <p>
        {t("feedbackInstruction", {
          PROGRAM_TYPE: t(`programTypeElative.${programItem.programType}`),
        })}
      </p>

      {!feedbackSent && (
        <>
          <FeedbackTextArea
            value={feedbackValue}
            onChange={handleFeedbackChange}
          />

          <ButtonWithMargin
            disabled={submitting}
            buttonStyle={ButtonStyle.SECONDARY}
            onClick={async () => await sendFeedbackEvent()}
          >
            {t("button.sendFeedback")}
          </ButtonWithMargin>
        </>
      )}

      {feedbackSent && (
        <SuccessMessage>{t("button.feedbackSent")}</SuccessMessage>
      )}
    </div>
  );
};

const FeedbackTextArea = styled(TextArea)`
  width: 100%;
  box-sizing: border-box;
`;

const SuccessMessage = styled.p`
  color: ${(props) => props.theme.textSecondary};
`;

const Title = styled.p`
  font-weight: 600;
`;

const ButtonWithMargin = styled(Button)`
  margin: 8px 0 16px 0;
`;
