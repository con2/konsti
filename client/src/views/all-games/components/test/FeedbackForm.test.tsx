import React from 'react';
import { shallow } from 'enzyme';
import { testGame } from 'test/test-data/testGame';
import { FeedbackForm, Props } from '../FeedbackForm';

describe('FeedbackForm', () => {
  it('should render correctly', () => {
    const props: Props = { game: testGame };
    const component = shallow(<FeedbackForm {...props} />);
    expect(component).toMatchSnapshot();
  });
});
