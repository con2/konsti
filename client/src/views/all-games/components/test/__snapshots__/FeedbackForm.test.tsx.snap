// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm should render correctly 1`] = `
<div
  className="feedback-form"
>
  <p
    className="bold"
  >
    feedbackTitle
  </p>
  <p>
    feedbackInstruction
  </p>
  <FeedbackForm__FeedbackTextarea
    onChange={[Function]}
    rows={4}
    value=""
  />
  <button
    disabled={false}
    onClick={[Function]}
  >
    button.sendFeedback
  </button>
</div>
`;
